File: modeling/R/dod.R

#setwd("/Users/apple/Desktop/dodpy/workspace/") #set working directory#KC#
setwd("C:/Users/tkriszti/Dropbox/rank and cohort life expectancy/date of death/dodpy/workspace")
library("splines")

sx_ages <- read.csv("Survival_ratio_Cohort_ages.csv")#


# ------------------------------- dist_sx function --------------------------- #
dist_sx <- function (CNTRY1,iSEX1,sx_date,sx_exact_age) {#KC#
  #find beginning of 5 yearly period for the sx_date
  sx_yr <- as.numeric(substr(sx_date,1,4))
  lowest_yr <- floor(sx_yr/5)*5 #lower limit of the five yearly period that includes sx_yr 
  
  lowest_age <- floor(sx_exact_age/5)*5
  selages <- match(paste("X",lowest_age-5,sep=""),names(sx_ages)):match("X125",names(sx_ages))#starting one step earlier
  
  #the diagonal is the cohort we are interested
  sx_mat  <- subset(sx_ages,region==CNTRY1 & sex==iSEX1 & Begin_prd >=(lowest_yr-5),select=selages) #extract a row corresponding to the time-period#KC#
  #select older cohort (for interpolation purpose)
  sx_mat_older  <- subset(sx_ages,region==CNTRY1 & sex==iSEX1 & Begin_prd >=(lowest_yr-10),select=c(selages)) #extract a row corresponding to the time-period#KC#
  #select younger cohort (for interpolation purpose)
  sx_mat_younger  <- subset(sx_ages,region==CNTRY1 & sex==iSEX1 & Begin_prd >=(lowest_yr),select=c(selages)) #extract a row corresponding to the time-period#KC#
  
  #dates for interpolation corresponding to the three periods 
  dates_x<- as.numeric(as.Date(c(paste(lowest_yr-5+3,1,1,sep="/"),paste(lowest_yr+3,1,1,sep="/"),paste(lowest_yr+5+3,1,1,sep="/")),"%Y/%m/%d"))
  #set place holder for calculation
  sx <- matrix(0,length(selages),7)
  colnames(sx) <- c("lower_age","pr0","pr1","pr2","pr_sx_date","death_percent","dth_pc_after_exact_age")
  #fill the first column with age (starting age of each five yearly age groups)
  sx[,1] <- seq((lowest_age-5),125,by=5)
  
  for ( i in 1:length(selages)) { #27 Sx values.. newborn, 0-4, ...., 125-129
    #fill the values
    sx[i,2] <- sx_mat_older[i,i]     
    sx[i,3] <- sx_mat[i,i]     
    sx[i,4] <- sx_mat_younger[i,i]     
    #interpolated to get SX for the sx_date
    sx[i,5] <- spline(sx[i,2:4]~dates_x,xout=as.numeric(sx_date))$y #   #interpSpline does not work with three values.. so used spline function here which also fits linearly
  } 
  
  #Calculate the % deaths starting at the lower age of the age-group where the sx_exact_age lies
  sx[1,6] <- 0
  sx[2,6] <- 100 #this is the starting population 
  for (i in 3:length(selages)) {
    sx[i,6] <- sx[i-1,6]*sx[i-1,5]
  } 
  
  #col7: % deaths
  for (i in 2:(length(selages)-1)) {
    sx[i,7] <- sx[i,6] - sx[i+1,6] 
  } 
  sx[i+1,7] <- sx[i+1,6] #everyone dies in this age-group
  
  #proportion of people who will die before the sx_exact_age
  die_earlier <- sx[2,7]*(sx_exact_age - lowest_age)/5
  sx[2,7] <-sx[2,7] - die_earlier #this is the proportion who died in this age-group after the sx_exact_age
  sx[,7] <- sx[,7] * 100/sum(sx[,7])
  sx[,1]<-sx[,1]+5
  
  sx[-1,c(1,7)]
} # function end

#Example a person who is aged 34.3 years on 2020-01-01
#this gives distribution of deaths (in percentage) of those 34.3 years old on 2020-01-01
#SX_DATE <- "2020-01-01"
SX_DATE <- Sys.Date()

SX_AGE <- 40 
deathsbyage <- dist_sx(CNTRY1="Nepal",iSEX1 = 1,sx_date=as.Date(SX_DATE),sx_exact_age=SX_AGE)
#this is the distribution of deaths by age
plot(deathsbyage,ylab = "Percent dying before Age", xlab="Age", type="b")

#finding percentiles
ispl <- interpSpline(cumsum(deathsbyage[,2])~deathsbyage[,1],bSpline = TRUE)
ispl
#function to calculate percentile
cal_percentile <- function (ispl=ispl,agex,perc) {
  (predict(ispl,agex)$y - perc)^2
}

#plotting...
plot(ispl, ylab = "Cumulative Percent dying before Age", xlab="Age")


#Median (e.g.)
PERC = 50
PERCAGE <- optimize(f=cal_percentile,interval = c(0:130),ispl=ispl,perc=PERC)$minimum
paste(PERC," percent of the population aged", SX_AGE, "years old who are alive on", SX_DATE,"will die before age",round(PERCAGE,1),"years",sep=" ")
points(x=PERCAGE,y=PERC)
lines(x=c(PERCAGE,PERCAGE),y=c(0,PERC))
lines(x=c(0,PERCAGE),y=c(PERC,PERC))
text(x=PERCAGE,y=0,paste("Median Age",round(PERCAGE,1),"yrs",sep=" "))


#First Quartile (25th percentile)
PERC = 25
PERCAGE <- optimize(f=cal_percentile,interval = c(0:130),ispl=ispl,perc=PERC)$minimum
paste(PERC," percent of the population aged", SX_AGE, "years old who are alive on", SX_DATE,"will die before age",round(PERCAGE,1),"years",sep=" ")
points(x=PERCAGE,y=PERC)

#####plotting declining percentage of live people#####

#finding percentiles

nrow(deathsbyage)  
alivebyage <- (cumsum(deathsbyage[nrow(deathsbyage):1,2]))[nrow(deathsbyage):1]
aliveage <- c(SX_AGE,deathsbyage[-(nrow(deathsbyage)),1])

#fitting spline 
ispl_live <- interpSpline(alivebyage~aliveage,bSpline = TRUE)

#plot
plot(ispl_live, ylab = "Cumulative Percent alive by Age", xlab="Age")


#Median (e.g.)
PERC = 50
PERCAGE <- optimize(f=cal_percentile,interval = c(0:130),ispl=ispl_live,perc=PERC)$minimum
paste(PERC," percent of the population aged", SX_AGE, "years old who are alive on", SX_DATE,"will die before age",round(PERCAGE,1),"years",sep=" ")
points(x=PERCAGE,y=PERC)
lines(x=c(PERCAGE,PERCAGE),y=c(0,PERC))
lines(x=c(0,PERCAGE),y=c(PERC,PERC))
text(x=PERCAGE,y=0,paste("Median Age",round(PERCAGE,1),"yrs",sep=" "))

File: modeling/python/dod.py
import csv
import os, sys
import pandas as pd
import numpy as np 
from datetime import datetime
import time
import math
from scipy import interpolate
#import matplotlib.pyplot as plt
from scipy.interpolate import splrep, InterpolatedUnivariateSpline, UnivariateSpline

# round data to nearest 5 year time window
def rounddown(x, base=5):
    return int(base * math.floor(float(x)/base))
# Get timedelta
def setInterpDate(x, offset):
    idate = datetime.strptime(str(x+offset+3)+"/1"+"/1", "%Y/%m/%d")
    return (idate - datetime(1970, 1, 1)).days
# generic days since epoch
def daysSinceEpoch(x, format="%Y-%m-%d"):
    idate = datetime.strptime(x,format)
    return (idate - datetime(1970, 1, 1)).days

# Import data
data = pd.read_csv('../../data/Survival_ratio_Cohort_ages.csv', header=0)

# --- wrap in function --- #
# get columns which correspond to the inputs
def dist_odata(cntry, sex, idate, iage):
    iyear = int(idate[0:4])
    flr_yr = rounddown(iyear, base=5)
    #print(iyear, flr_yr)

    # get closest age in 5 year windows
    flr_age = rounddown(iage, base=5)
    #print(flr_age)

    # Get the age cohort
    if flr_age>=5:
        cohort_st = list(data.columns).index("X"+str(flr_age-5))
    else:
        cohort_st = 4
    cohort_end = len(data.columns)
    #cohort = data.ix[:,cohort_st:cohort_end]
    cohort = data.loc[(data.region==cntry) & (data.sex==sex) & (data.Begin_prd >=(flr_yr-5))].ix[:,cohort_st:cohort_end]
    #print(cohort.columns)

    # get older and younger cohort
    cohort_old = data.loc[(data.region==cntry) & (data.sex==sex) & (data.Begin_prd >=(flr_yr-10))].ix[:,cohort_st:cohort_end]
    cohort_young = data.loc[(data.region==cntry) & (data.sex==sex) & (data.Begin_prd >=(flr_yr))].ix[:,cohort_st:cohort_end]

    # get dates for the jan 1st for 3 years --> then to Unix timestamp
    dates = [setInterpDate(flr_yr, -5), setInterpDate(flr_yr,0), setInterpDate(flr_yr,+5)]
    #print(dates)

    # make the output datatable
    temp = np.zeros(shape=(len(cohort.columns),7))
    odata = pd.DataFrame(temp, columns=["lower_age","pr0","pr1","pr2","pr_sx_date","death_percent","dth_pc_after_exact_age"])

    # fill in with existing values
    if iage>=5:
        odata['lower_age'] =np.arange(iage-5, 130, 5)
    else:
        odata['lower_age'] =np.arange(0, 130, 5)
    odata['pr0'] = np.matrix(cohort_old).diagonal().T
    odata['pr1'] = np.matrix(cohort).diagonal().T
    odata['pr2'] = np.matrix(cohort_young).diagonal().T

    # Interpolate for the input date (idate)
    odata["pr_sx_date"] = np.array([InterpolatedUnivariateSpline(dates,list(odata.ix[i,1:4]),k=2)(daysSinceEpoch(idate)) for i in np.arange(0, len(cohort.columns),1)])


    clen = len(odata)
    # calc the % deaths
    # odata["death_percent"][0] = 0
    odata["death_percent"][1] = 100
    #odata["death_percent"][2:clen] = [((odata["death_percent"][i-1])*(odata["pr_sx_date"][i-1])) for i in np.arange(2,clen,1) ]
    for i in np.arange(2,clen,1):
        odata["death_percent"][i] = odata["death_percent"][i-1]*odata["pr_sx_date"][i-1]


    # percentage deaths
    #odata["dth_pc_after_exact_age"][1:clen-1] = [ (odata["death_percent"][i] - odata["death_percent"][i+1])  for i in  np.arange(1,clen-1, 1)]
    for i in np.arange(1,clen-1,1):
        odata["dth_pc_after_exact_age"][i] = odata["death_percent"][i] - odata["death_percent"][i+1]

    odata["dth_pc_after_exact_age"][clen-1] = odata["death_percent"][clen-1]
    #print(odata)

    # proportion of people who will die before iage
    beforeDod = odata["dth_pc_after_exact_age"][1] * (iage - flr_age)/5
    odata["dth_pc_after_exact_age"][1] = odata["dth_pc_after_exact_age"][1] - beforeDod
    odata["dth_pc_after_exact_age"] = odata["dth_pc_after_exact_age"]* 100/odata["dth_pc_after_exact_age"].sum()


    # add 5 to each of the "ages"
    if iage>=5:
        odata["lower_age"] = odata["lower_age"]+5
    else:
        odata["lower_age"] = odata["lower_age"]+iage

    output = odata.ix[1:clen-1,['lower_age', 'dth_pc_after_exact_age']]
    return output
# --- function end --- #

# data input
cntry = "Nepal"
sex = 1 #1=Man, 2=Woman
idate = time.strftime('%Y-%m-%d')
iage =0

#print(cntry, sex, idate, iage)
output = dist_odata(cntry, sex, idate, iage)
print output

#plt.plot(output['lower_age'], output['dth_pc_after_exact_age'])
# plt.show()

# spl = InterpolatedUnivariateSpline(output['lower_age'], output['dth_pc_after_exact_age'])
# yrs = np.arange(output['lower_age'][0],output['lower_age'][len(output)-1])
# plt.plot(yrs,abs(spl(yrs)))

# --- part 2 --- #
# find percentiles
#f = InterpolatedUnivariateSpline(output['lower_age'], output['dth_pc_after_exact_age'].cumsum())
#print f.get_coeffs()




'''
# time for list comprehension!
f = np.array([interpolate.interp1d(dates, odata.ix[i,1:4])(daysSinceEpoch(idate)) for i in np.arange(0, len(cohort.columns),1)])
print f

# test interpolation
test = np.arange(1, 4, 1)
f = interpolate.interp1d(dates, test)
print f(daysSinceEpoch(idate))

f= []
for i in np.arange(1,len(odata["death_percent"]),1):
    f.append((odata["death_percent"][i]-1)*(odata["pr_sx_date"][i]-1))
print f[::-1]


'''